### PR TITLE
[fix](distributed) Isolate process-local Flight service lifecycle

### LINKED ISSUE #73: [P1][Bug] Give the process-local Flight service explicit ownership and exchange isolation
## Problem

The worker process intentionally needs one shared Arrow Flight server, but the current singleton has implicit ownership and mixes process-level transport configuration with exchange/session storage configuration.

`EnsureLocalFlightServerStarted()` treats `node_id` equality as sufficient for reuse. A later exchange can therefore silently inherit an earlier bind host, port, and `local_dirs`. `FlightExchangeManager::Shutdown()` is empty, so no explicit owner can drain the server, join its thread, and wait for port release.

The data namespace is also process-global: tickets and `ShuffleCacheRegistry` identify data primarily by stage ID. Independent sessions with overlapping stage IDs can collide, and an old ticket has no server generation or exchange generation with which to become stale.

### PR DESCRIPTION
**Pull request by kaka11chen**
## Summary

- make the worker process own one Flight service with immutable bind configuration, explicit shutdown, thread join, and port release
- publish committed shuffle attempts in a process-local catalog under unique exchange-instance identities and exact query ownership
- return the native completed sink descriptor so the worker Flight epoch reaches the coordinator
- serialize sink network and local-directory settings from the owning Flight session instead of rereading process environment
- fence query teardown before task cancellation, track the real native thread lifetime, and reject late cache publication after close
- keep deferred cleanup in an explicit pending queue; report, retry, and drain reader-held or failed cleanup before teardown completes
- carry the server epoch through sink/source metadata and use a strict six-field `v1` ticket (`epoch`, exchange instance, node, attempt, partition)
- resolve remote reads only through live catalog entries, rejecting stale epochs, released attempts, and unpublished manifests

## Intentional wire behavior

The ticket version remains `v1`, but its schema is replaced in place. The previous four-field `v1` shape is rejected. This PR intentionally adds no compatibility parser and no server-side directory or manifest fallback.

## Validation

- `scripts/format workspace --changed` — passed
- incremental PEP 517 native build/install — passed
- `scripts/run_native_tests.sh '[distributed][exchange]'` — 38 cases, 459 assertions passed
- native serialization/exchange filter — 7 cases, 115 assertions passed
- Flight-disabled C++20 `duckdb_static` build — 270/270 targets passed
- affected Python suite — 559 passed, 4 deselected
- `scripts/run_release_tests.sh` — 278 passed, 1 skipped, then 5 Ray integration tests passed
- `scripts/run_fast_tests.sh` non-Ray phase — 5,391 passed, 27 skipped, 8 xfailed, 1 xpassed
- `scripts/run_fast_tests.sh` shared-Ray phase — 78 passed, 6 skipped
- `scripts/run_fast_tests.sh` owner-Ray phase — 6 passed, 3 environment-dependent skips, including a passing non-contiguous CUDA visibility integration test

Fixes #73

### CODE REVIEW

File: external/duckdb/src/execution/physical_operator.cpp [hunk 1]
```diff
@@ -1165,6 +1165,7 @@ unique_ptr<PhysicalOperator> PhysicalOperator::DeserializeOperatorData(Deseriali
 		    deserializer.ReadPropertyWithExplicitDefault<vector<string>>(114, "range_boundaries", {});
 		auto range_order_modifiers =
 		    deserializer.ReadPropertyWithExplicitDefault<vector<string>>(115, "range_order_modifiers", {});
+		sink_handle.flight_server_epoch = deserializer.ReadProperty<string>(116, "flight_server_epoch");
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply external/duckdb/src/execution/physical_operator.cpp:1168]: This is intentional for issue #73: the v1 wire schema is replaced in place and older serialized operators must be rejected. We explicitly do not add a compatibility default or fallback, so the required read is the intended behavior.

File: external/duckdb/src/execution/distributed/plan/exchange_sink_instance_task.cpp [hunk 2]
```diff
@@ -64,6 +65,7 @@ ExchangeSinkInstanceTaskDescriptor ExchangeSinkInstanceTaskDescriptor::Deseriali
 	    deserializer.ReadPropertyWithExplicitDefault<string>(3, "output_location", "");
 	result.sink_instance.output_partition_count =
 	    deserializer.ReadPropertyWithDefault<idx_t>(4, "output_partition_count");
+	result.sink_instance.flight_server_epoch = deserializer.ReadProperty<string>(5, "flight_server_epoch");
 	return result;
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply external/duckdb/src/execution/distributed/plan/exchange_sink_instance_task.cpp:125]: This is intentional for issue #73: the v1 descriptor schema is replaced in place and older sink descriptors must be rejected. No compatibility default or fallback is part of this change.

File: external/duckdb/src/execution/distributed/exchange/flight_server.cpp [hunk 3]
```diff
@@ -358,11 +313,7 @@ FlightServer::~FlightServer() {
...
 	if (impl_ && impl_->server()) {
-		auto status = impl_->server()->Shutdown();
-		if (!status.ok()) {
-			server_thread_.detach();
-			return;
-		}
+		(void)impl_->server()->Shutdown();
 	}
 	server_thread_.join();
 }
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply external/duckdb/src/execution/distributed/exchange/flight_server.cpp:320]: Detaching is not lifetime-safe here: the Serve/DoGet thread retains access to the FlightServer and its catalog, so detaching after a failed Shutdown would allow use-after-free once destruction continues. Explicit worker teardown surfaces Stop/Shutdown failures while the owner is still alive; at final destruction, joining is the only safe ownership policy.

File: external/duckdb/src/execution/distributed/plan/exchange_source_task.cpp [hunk 4]
```diff
@@ -95,6 +96,7 @@ ExchangeSourceTaskDescriptor ExchangeSourceTaskDescriptor::Deserialize(Deseriali
 				handle.files.push_back(std::move(file));
 			});
 			handle.attempt_id = obj.ReadPropertyWithExplicitDefault<idx_t>(5, "attempt_id", 0);
+			handle.flight_server_epoch = obj.ReadProperty<string>(6, "flight_server_epoch");
 		});
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply external/duckdb/src/execution/distributed/plan/exchange_source_task.cpp:100]: This is intentional for issue #73: the v1 descriptor schema is replaced in place and older source descriptors must be rejected. No compatibility default or fallback is part of this change.

**Comment by kaka11chen** [comment]
@codex review

File: src/duckdb_py/ray/ray_module.cpp [hunk 5]
```diff
@@ -913,6 +1003,13 @@ void register_ray_bindings(py::module_ &mod) {
 					        exchange_sink_instance_task.sink_instance.output_location =
 					            py::str(d["attempt_path"]).cast<string>();
 				        }
+				        if (d.contains("flight_server_epoch")) {
+					        exchange_sink_instance_task.sink_instance.flight_server_epoch =
+					            py::str(d["flight_server_epoch"]).cast<string>();
+				        }
+				        if (d.contains("query_id")) {
+					        exchange_sink_instance_task.sink_instance.query_id = py::str(d["query_id"]).cast<string>();
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply src/duckdb_py/ray/ray_module.cpp:1015]: Fixed in bd87af00bf. Both Python-to-C++ exchange sink instance parsers now read query_id from sink_handle while retaining the existing top-level form. Regression coverage uses ExchangeSinkInstanceHandle.to_dict() through both native plan application and Ray task-result materialization.

File: duckdb/runners/fte/fte_worker_runtime.py [hunk 6]
```diff
@@ -1471,3 +1476,35 @@ async def drop_query(self, query_id: str) -> dict[str, int]:
...
+                        canceled += 1
+                    execution.release_result(reason="worker_shutdown")
+                except BaseException as exc:
+                    errors.append((key, exc))
+                    continue
+                self.tasks.pop(key, None)
+                self.running_tasks.discard(key)
+                removed += 1
+            self.query_tasks.clear()
+            self.queued_tasks.clear()
```
[the remark this answers is not in the document]
Comment (kaka11chen) [reply duckdb/runners/fte/fte_worker_runtime.py]: Fixed in bd87af00bf. shutdown now clears queued_tasks immediately after setting _shutdown, and _drain_queue is serialized by the lifecycle lock with an explicit shutdown guard. A deterministic regression test simulates a completion callback during shutdown and verifies no queued task is admitted.

**Comment by kaka11chen** [comment]
@codex review

**Comment by kaka11chen** [comment]
@codex review

**Review by QuakeWang — APPROVED** [review]
LGTM
